rfcs/message-first-architecture/RFC.md

```diff
@@ -0,0 +1,211 @@
+})
+```
+
+```ts
+// For comparison, here is the current API:
+
+const refResource = resources.find((resource) => resource.languageTag.name === "en-US")
+const message = query(refResource).get({
+	id: "hello.login",
+})
+
+// -------------------
+
+// this proposal
+const message = messages.get({ id: "hello.login", languageTag: "en-US" })
+```
+
+#### 3. `config.saveMessages()`
+
+- A plugin can save messages anywhere (e.g. resources files, remote APIs, embedded databases).
+
+```ts
+await config.saveMessages(messages)
+```
+
+### Plugins
+
+- nothing changes for plugins except that they need to expose `saveMessages()` and `getMessages()`.
+
+### Linting
+
+- No more resources.
+  -- Lint "messages" makes much more sense than linting "resources".
+  -- Linting "messages" is what our lint system is about.
+  -- Linting "resources" is the job of eslint, prettier, etc.
+
+- Simpler internal structure.
+
+#### Linting application flow
+
+```ts
+const [lints, errorsDuringLinting] = await lint({ messages, config })
+
+lints.reports.filter((report) => report.level === "error")
+```
+
+```
+[
+  {
+    result: "Missing translation for message 'hello' in language 'fr'."
+    ...
+  },
+  {
+    result: "Missing translation for message 'hello' in language 'de-DE'.",
+    ...
+  },
+]
+
+```
+
+#### Writing lint rules
+
+- Simple, plain JS. Simple to explain and maintain linting system.
+- Performance optimized via the `messages` object.
+
+##### Currently
+
+```ts
+export const missingMessage = createLintRule({
+	id: "inlang.missingMessage",
+	setup: ({ report }) => {
+		let targetLanguage: LanguageTag["name"] | undefined
+		return {
+			visitors: {
+				Resource: ({ target }) => {
+					// we need to derive the target language from the resource
+					// because the message is missing.
+					targetLanguage = target?.languageTag.name
+				},
+				Message: ({ target, reference }) => {
+					if (target === undefined && reference) {
+						report({
+							node: reference,
+							message: `Message with id '${reference.id.name}' is missing for '${targetLanguage}'.`,
+						})
+						return {
+							ruleId: reference.id.name,
+							languageTag: targetLanguage,
+							message: `Message with id '${reference.id.name}' is missing for '${targetLanguage}'.`,
+						}
+					}
+				},
+			},
+		}
+	},
+})
+```
+
+##### Proposal
+
+> Note: This is a sketch. The API design should foster a discussion about the general direction.
+
+- No more `Resource` nodes. Just `Message` linting.
+- No more `target` and `reference` nodes. Just a `message` and all the `messages`.
+- No more `"skip"` concept because performance is optimized via the `messages` object.
+- The mental model is simple: "I have a message and I can query all the messages to derive is something is wrong"
+
+```ts
+export const missingMessage = createLintRule({
+	id: "inlang.missingMessage",
+	message: ({ message, messages, config }) => {
+		if (message.languageTag === config.referenceLanguageTag) {
```

Comment:
```suggestion
		if (message.languageTag !== config.referenceLanguageTag) {
```

You want to iterate over the messages of the referece language for this lint rule